File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/domain/events/ExampleCreatedEvent.java
package com.oddly.ddd.domain.events;

import com.oddly.ddd.domain.events.infra.BaseDomainEvent;

/**
 * Domain event for when an example is created.
 * Domain events:
 * - Have Event suffix
 * - Follow pattern: {Object}{Action}Event
 * - Are immutable (all fields final)
 * - Live in /domain/events/
 */
public class ExampleCreatedEvent extends BaseDomainEvent {
    private final String m_exampleId;
    private final String m_name;
    private final String m_ownerId;

    public ExampleCreatedEvent(String p_exampleId, String p_name, String p_ownerId, String p_correlationId) {
        super(p_correlationId);
        this.m_exampleId = p_exampleId;
        this.m_name = p_name;
        this.m_ownerId = p_ownerId;
    }

    @Override
    public String getEventType() {
        return "example.created";
    }

    public String getExampleId() {
        return m_exampleId;
    }

    public String getName() {
        return m_name;
    }

    public String getOwnerId() {
        return m_ownerId;
    }
}

File: infrastructures/ddd/csharp/src/Domain/Events/ExampleCreatedEvent.cs
namespace OddlyDdd.Domain.Events
{
    /// <summary>
    /// Domain event for when an example is created.
    /// Domain events MUST:
    /// - Suffix with "Event"
    /// - Follow pattern: {Object}{Action}Event
    /// - Be immutable (init-only properties)
    /// - Live in /Domain/Events/
    /// - Include eventId, timestamp, and correlationId (from base)
    /// 
    /// Used for:
    /// - Asynchronous subdomain-to-subdomain communication
    /// - ONLY way to communicate between subdomains (NO HTTP calls)
    /// 
    /// Topic naming: {subdomain}.created (e.g., "example.created")
    /// </summary>
    public class ExampleCreatedEvent : BaseDomainEvent
    {
        /// <summary>
        /// The ID of the created example
        /// </summary>
        public string ExampleId { get; init; }

        /// <summary>
        /// The name of the created example
        /// </summary>
        public string Name { get; init; }

        /// <summary>
        /// The owner ID of the created example
        /// </summary>
        public string OwnerId { get; init; }

        public ExampleCreatedEvent(
            string p_exampleId,
            string p_name,
            string p_ownerId,
            string p_correlationId
        ) : base(p_correlationId)
        {
            ExampleId = p_exampleId;
            Name = p_name;
            OwnerId = p_ownerId;
        }
    }
}
